src/linalg/muladd.jl

```diff
@@ -47,54 +35,83 @@ length(M::MulAdd) = prod(size(M))
 size(M::MulAdd) = length.(axes(M))
 axes(M::MulAdd) = axes(M.C)
 
+similar(M::MulAdd, ::Type{T}, axes) where {T,N} = Array{T}(undef, length.(axes))
+similar(M::MulAdd, ::Type{T}) where T = similar(M, T, axes(M))
+similar(M::MulAdd) = similar(M, eltype(M))
+
+function instantiate(M::MulAdd)
+    @boundscheck check_mul_axes(M.α, M.A, M.B)
+    @boundscheck check_mul_axes(M.β, M.C)
+    @boundscheck axes(M.A,1) == axes(M.C,1) || throw(DimensionMismatch("First axis of A, $(axes(A,1)), and first axis of C, $(axes(C,1)) must match"))
+    @boundscheck axes(M.B,2) == axes(M.C,2) || throw(DimensionMismatch("Second axis of B, $(axes(B,2)), and second axis of C, $(axes(C,2)) must match"))
+    M
+end
+
+struct Lmul{StyleA, StyleB, TypeA, TypeB}
+    A::TypeA
+    B::TypeB
+end
+
+Lmul(A::TypeA, B::TypeB) where {TypeA,TypeB} = Lmul{typeof(MemoryLayout(TypeA)),typeof(MemoryLayout(TypeB)),TypeA,TypeB}(A,B)
+
+
 BroadcastStyle(::Type{<:MulAdd}) = ApplyBroadcastStyle()
 BroadcastStyle(::Type{<:Lmul}) = ApplyBroadcastStyle()
 
 struct LmulStyle <: AbstractArrayApplyStyle end
-abstract type AbstractMulAddStyle <: AbstractArrayApplyStyle end
-struct MulAddStyle <: AbstractMulAddStyle end
-
+struct MulAddStyle <: AbstractArrayApplyStyle end
+struct ScalarMulStyle <: ApplyStyle end
+struct IdentityMulStyle <: AbstractArrayApplyStyle end
+
+combine_mul_styles() = MulAddStyle()
+combine_mul_styles(_) = IdentityMulStyle()
+combine_mul_styles(::ScalarLayout) = ScalarMulStyle()
+combine_mul_styles(c1, c2) = result_mul_style(combine_mul_styles(c1), combine_mul_styles(c2))
+@inline combine_mul_styles(c1, c2, cs...) = result_mul_style(combine_mul_styles(c1), combine_mul_styles(c2, cs...))
```

Comment:
I just copied Base. I'm inclined to leave it as it works, and things like `mapfoldl` can be delicate inference wise (you may have noticed I had to change a `mapreduce` call which worked in 1.2 but didn't in 1.0/1.1)

Reply:
In many projects I just define

```julia
@inline foldlargs(op, x) = x
@inline foldlargs(op, x1, x2, xs...) = foldlargs(op, op(x1, x2), xs...)
```

and use it instead of `Base.foldl`. I personally think using common term like fold makes it easier to emphasize the pattern. But if you want to keep it as-is it also makes sense too.

Reply:
Both suggestions are reasonable, but could easily be done as a separate PR. I don’t have that much steam left with the remaining Julia v1.0 bugs, so maybe I’ll hold off on it?

Reply:
Right, it doesn't have to be done in a big rewrite PR like this.